### ISSUE TITLE
SQlite INTEGER should map to i64 not i32

### ISSUE THREAD

**Issue by data-retriever** [SeaQL/sea-orm#1067]
From SQlite's [documentation](https://www.sqlite.org/datatype3.html):
```
INTEGER. The value is a signed integer, stored in 0, 1, 2, 3, 4, 6, or 8 bytes depending on the magnitude of the value.
```

But `sea-orm-cli generate entity` generates a i32

**Comment by data-retriever** [comment]
Ok, after some random testing I see that if the type is BIGINT but not (BIGINTEGER) on the sql statement then the generated type is i64.

So, SQL -> Rust:
field INTEGER -> field: i32
field BIGINT -> field: i64
field BIGINTEGER -> field: Vec<u8>

This is really surprising. Specially coming from sqlx that would handle INTEGER as i64

**Comment by billy1624** [comment]
Hey @data-retriever, welcome! Nice icon lolll

Note that in SQLite it only has `INTEGER`, `TEXT`, `BLOB`, `REAL` and `NUMERIC` datatype. So, for example `BIGINT` will be affinized as `INTEGER` and `BIGINTEGER` will be affinized as `NUMERIC`.

For CLI codegen, the mapping from SQLite datatype to Rust type is providing by SeaSchema at here
- https://github.com/SeaQL/sea-schema/blob/c5a3b86fb7634b281d8a599ada3193935db87716/src/sqlite/def/types.rs#L43-L77

**Comment by anshulxyz** [comment]
@data-retriever About SQLite type affinity refer https://www.sqlite.org/datatype3.html#determination_of_column_affinity

**Comment by billy1624** [comment]
That's correct, SQLite did support up to i64. But I think for most use cases i32 is enough and considering i64 is much more costly than i32. So, I feel i32 is a good default. Demanding user can change the integer field to i64 easily.

Input? @tyt2y3

**Comment by data-retriever** [comment]
> That's correct, SQLite did support up to i64. But I think for most use cases i32 is enough and considering i64 is much more costly than i32. So, I feel i32 is a good default. Demanding user can change the integer field to i64 easily.
> 
> Input? @tyt2y3

Perhaps what's missing is some comment in the documentation. It's not clear at all that using BIGINT is the way to go.

**Comment by billy1624** [comment]
> Ok, after some random testing I see that if the type is BIGINT but not (BIGINTEGER) on the sql statement then the generated type is i64.
> 
> So, SQL -> Rust:
> 
> ```
> field INTEGER -> field: i32
> field BIGINT -> field: i64
> field BIGINTEGER -> field: Vec<u8>
> ```
> 
> This is really surprising. Specially coming from sqlx that would handle INTEGER as i64.
> 
> Also, why is BIGINT different from BIGINTEGER?

Short answer, yes, `INT` / `BIGINT` is the way to go.

Circle back to the original question. According to https://www.sqlite.org/datatype3.html, `INT`, `BIGINT`...etc will be treated as `INT` but `BIGINTEGER` isn't mention so that's why SeaSchema parse it as `BLOB` column type.

**Comment by AngelOnFira** [comment]
Another inconsistency that I've run into is when

`creating migration -> running migration -> creating entities`

I'm using Sqlite for development of a project, and I've written a migration as such:
```rust
manager
    .create_table(
        Table::create()
            .table(Channel::Table)
            ...
            .col(ColumnDef::new(Channel::DiscordId).big_integer().not_null())
            ...
            .to_owned(),
    )
    .await?;
```
In this case, running the migration against a Sqlite DB will change that `big_integer` call to an `INTEGER` in Sqlite, as noted above. So when entities are generated with `sea-orm-cli`, the model becomes:
```rust
pub struct Model {
    ...
    pub discord_id: i32,
    ...
}
```
As far as I know, there's no way for me to annotate somewhere that I want it to come out as an i64 (which makes sense, I don't know of anywhere that would make sense to go). Since I'm generating entities, it's not as easy as being the "demanding user" that will go through and switch things out all the time.

For a sensible default, I feel it makes more sense to be an `i64` than an `i32`. It's ambiguous since data is lost when extracting from Sqlite, so I understand why `i32` seems fine, since saving space is good. However, generating entities from an existing Sqlite DB that makes full use of 8 bytes of data in an INTEGER field would mean data is lost as the default behaviour.

**Comment by AngelOnFira** [comment]
I threw together a repo that can be found here: https://github.com/AngelOnFira/sea-orm-sqlite-integer-issue

Essentially, I generated entities from a Sqlite database with an INTEGER column that held a number larger than an `i32`, and when I read the number in using SeaORM, it didn't retrieve the correct number.

tl;dr `999_999_999_999_999` came out as `-1530494977`. I think this potential behaviour should be considered incorrect, and hopefully, this can motivate changing the default :)

**Comment by data-retriever** [comment]
> For a sensible default, I feel it makes more sense to be an `i64` than an `i32`. It's ambiguous since data is lost when extracting from Sqlite, so I understand why `i32` seems fine, since saving space is good. However, generating entities from an existing Sqlite DB that makes full use of 8 bytes of data in an INTEGER field would mean data is lost as the default behaviour.

I agree 100%. The default behavior is very surprising for someone that was already working with sqlite databases before using seaorm.

But also it seems there is another issue. It seems that on the DSL there is only `.big_integer()`, no `.big_int()` or similar. If using SQL the workaround is to use `field bigint` instead of `field integer` or `field biginteger`. So how would someone create i64 fields with the DSL?

**Comment by tyt2y3** [comment]
Thank you everyone for bringing this up.

I think we should map sea-query's `BigInteger` to `bigint`
https://github.com/SeaQL/sea-query/blob/485b0872669720d33f0f347ba7211799321f4c1b/src/backend/sqlite/table.rs#L55

This way, it will be picked up correctly by sea-schema thus completing the loop.

But mapping `int` to `i32` and `bigint` to `i64` is by design, in an effort to 'try' to align the behaviour with MySQL / Postgres, in which `int` is 32 bit.

**Comment by microidea** [comment]
This is my SQLite table:
```sql
CREATE TABLE IF NOT EXISTS "user"
(
    id          INTEGER not null
        constraint user_pk
            primary key,
    data        TEXT    not null,
    create_time INTEGER not null,
    update_time INTEGER
)
    strict
```
and I use `sea-orm-cli generate entity` to generate entity file, the result is:
```rust
//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.4

use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "user")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    pub id: i32,
    #[sea_orm(column_type = "Text")]
    pub data: String,
    pub create_time: i32,
    pub update_time: Option<i32>,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {}

impl ActiveModelBehavior for ActiveModel {}
```

How can I change i32 to i64 in the entity file? Is there any easy way to do this?
I don't want to do it by manually editing the file, since it will be overwrite every time I run `sea-orm-cli generate entity`.
Thanks.

**Comment by nicka101** [comment]
> But mapping int to i32 and bigint to i64 is by design, to align the behaviour with MySQL / Postgres, in which int is 32 bit.

Feature parity between databases might be nice, but fundamentally breaking type handling by forcing a 64 bit value into 32 bits is definitely not... this issue should be reopened.

Sqlite integer types go up to 8 bytes, and are signed by default, and thus `sea-orm-cli` should map them as `i64`.

Failure to do this results in any value above the 32 bit threshold being horribly broken, and could potentially cause data loss.

Currently there is no way for sea-orm to correctly map ROWID columns to their actual size (64 bit), they are always mapped as `i32`, which is **dangerously** incorrect. From the Sqlite docs:

> In SQLite, a column with type INTEGER PRIMARY KEY is an alias for the [ROWID](https://sqlite.org/lang_createtable.html#rowid) (except in [WITHOUT ROWID](https://sqlite.org/withoutrowid.html) tables) which is always a 64-bit signed integer.

At the very least a flag should be added to `sea-orm-cli` for it to use correctly sized types.
Currently the best option available for any ROWID column is to manually change the generated type to `i64`, which quickly becomes tediously repetitive.
My only current DB target is sqlite, so I have no need for parity between the database engines, but I do require consistently correct data handling
